mav_trajectory_generation/include/mav_trajectory_generation/impl/polynomial_optimization_nonlinear_impl.h

```diff
@@ -230,6 +234,351 @@ int PolynomialOptimizationNonLinear<_N>::optimizeTimeAndFreeConstraints() {
+//              << " total cost: " << w_d*J_d+w_t*J_t+w_sc*J_sc
+//              << " gradient norm: " << grad.norm()
+//              << std::endl;
+//
+//    std::cout << "[GD] i: " << i << " grad: " << grad.transpose()
+//              << std::endl;
+//    std::cout << "[GD] i: " << i << " increment: " << increment.transpose()
+//              << std::endl;
+//
+//    std::cout << "grad_d | grad_sc: "
+//              << std::endl;
+//    for (int k = 0; k < dim; ++k) {
+//      for (int n = 0; n < n_free_constraints; ++n) {
+//        std::cout << k << " " << n << ": " << grad_d[k][n] << " | "
+//                  << grad_sc[k][n] << std::endl;
+//      }
+//      std::cout << std::endl;
+//    }
+//    std::cout << std::endl;
+//
+//    std::cout << "grad_t: " << std::endl;
+//    for (int i = 0; i < n_segments; ++i) {
+//      std::cout << grad_t[i] << " | ";
+//    }
+//    std::cout << std::endl << std::endl;
+
+    // Update the parameters.
+    x += increment;
+    // Check that segment times are > 0.1s
+    for (int n = 0; n < n_segments; ++n) {
+      x[n] = x[n] <= 0.1 ? 0.1 : x[n];
+    }
+
+    // Set new segment times and new free constraints
+    std::vector<double> segment_times_new;
+    segment_times_new.reserve(n_segments);
+    for (size_t i = 0; i < n_segments; ++i) {
+      segment_times_new.push_back(x[i]);
+    }
+    std::vector<Eigen::VectorXd> d_p_vec_new;
+    d_p_vec_new.resize(dim, Eigen::VectorXd::Zero(n_free_constraints));
+    for (int k = 0; k < dim; ++k) {
+      d_p_vec_new[k] = x.block(n_segments+k*n_free_constraints, 0,
+                               n_free_constraints, 1);
+    }
+
+    // Update segement times and free constraints
+    poly_opt_.updateSegmentTimes(segment_times_new);
+    poly_opt_.setFreeConstraints(d_p_vec_new);
+    poly_opt_.solveLinear(); // TODO: needed?
+  }
+
+  // Print all parameter
+//  std::cout << "[GD Original]: " << x_orig.transpose() << std::endl;
+//  std::cout << "[GD Solution]: " << x.transpose() << std::endl;
+  // Print only segment times
+  std::cout << "[GD Original]: " << x_orig.block(0,0,n_segments,1).transpose()
+            << std::endl;
+  std::cout << "[GD Solution]: " << x.block(0,0,n_segments,1).transpose()
+            << std::endl;
+  std::cout << "[GD Trajectory Time] Before: "
+            << x_orig.block(0,0,n_segments,1).sum()
+            << " | After: " << x.block(0,0,n_segments,1).sum() << std::endl;
+
+  return nlopt::SUCCESS;
+}
+
+template <int _N>
+double PolynomialOptimizationNonLinear<_N>::getCostAndGradientTimeForward(
+        std::vector<double>* gradients) {
+
+  // Weighting terms for different costs
+  const double w_d = 0.1;
+  const double w_t = 5.0;
+  const double w_sc = 1.0;
+
+  // Retrieve the current segment times
+  std::vector<double> segment_times;
+  poly_opt_.getSegmentTimes(&segment_times);
+
+  // Calculate current cost
+  // TODO: parse from outside?
+  const double J_d = 2*poly_opt_.computeCost();// TODO: *2 necessary?
```

Comment:
Since the cost in poly_opt_.computeCost() are calculated as follows: cost = 0.5 * c^T * Q * c
I guess it doesnt change much but i thought it would be nice to have the original value?